src/realm/object-store/impl/deep_change_checker.hpp

```diff
@@ -0,0 +1,249 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2021 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include <realm/object-store/object_changeset.hpp>
+#include <realm/object-store/impl/collection_change_builder.hpp>
+#include <realm/object-store/util/checked_mutex.hpp>
+
+#include <realm/util/assert.hpp>
+#include <realm/version_id.hpp>
+#include <realm/keys.hpp>
+#include <realm/table_ref.hpp>
+
+#include <array>
+#include <atomic>
+#include <exception>
+#include <functional>
+#include <mutex>
+#include <unordered_map>
+#include <unordered_set>
+
+namespace realm {
+class Realm;
+class Transaction;
+
+using KeyPath = std::vector<std::pair<TableKey, ColKey>>;
+using KeyPathArray = std::vector<KeyPath>;
+
+namespace _impl {
+class RealmCoordinator;
+
+struct ListChangeInfo {
+    TableKey table_key;
+    int64_t row_key;
+    int64_t col_key;
+    CollectionChangeBuilder* changes;
+};
+
+// FIXME: this should be in core
+using TableKeyType = decltype(TableKey::value);
+using ObjKeyType = decltype(ObjKey::value);
+
+// A collection of all changes to all tables which we use to check against when the `DeepChangeChecker`.
+struct TransactionChangeInfo {
+    std::vector<ListChangeInfo> lists;
+    std::unordered_map<TableKeyType, ObjectChangeSet> tables;
+    bool track_all;
+    bool schema_changed;
+};
+
+/**
+ * The `DeepChangeChecker` serves two purposes:
+ * - Given an initial `Table` and an optional `KeyPathArray` it find all tables related to that initial table.
+ *   A `RelatedTable` is a `Table` that can be reached via a link from another `Table`.
+ * - The `DeepChangeChecker` also offers a way to check if a specific `ObjKey` was changed.
+ */
+class DeepChangeChecker {
+public:
+    struct OutgoingLink {
+        int64_t col_key;
```

Comment:
Is there a reason we want to store the raw int64_t here instead of the ColKey type which is just an int64_t ?

Reply:
It has been that way already. We do have several locations where we just save the raw value.
There are even definitions like:
```
using TableKeyType = decltype(TableKey::value);
using ObjKeyType = decltype(ObjKey::value);
```

I'd like the idea of only using the actual struct / type (TableKey, ColKey, ObjKey, etc.) but maybe someone else can explain this better. :)

Reply:
This code was written before ColKey etc. existed and the core 6 port basically involved doing the bare minimum to get things to compile and pass tests.